File: Animal Age/DeviceViewController.m
//
//  DeviceViewController.m
//  Animal Age Converter
//
//  Created by Jon on 3/27/17.
//  Copyright © 2017 Jon Brown. All rights reserved.
//

#import "DeviceViewController.h"
#import "AppDelegate.h"
#import "DeviceDetailViewController.h"

@interface DeviceViewController ()
@property (strong) NSMutableArray *devices;

@end

@implementation DeviceViewController

-(NSManagedObjectContext *)managedObjectContext{
    
    NSManagedObjectContext *context = nil;
    id delegate = (AppDelegate *)[UIApplication sharedApplication].delegate;
    // call "persistentContainer" not "managedObjectContext"
    if( [delegate performSelector:@selector(persistentContainer)] ){
        // call viewContext from persistentContainer not "managedObjectContext"
        context = [[delegate persistentContainer] viewContext];
    }
    
    return context;
}

- (void)viewDidLoad {
    
    [super viewDidLoad];
    self.title = @"Age Tracker";
}

- (void) viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    // You code here to update the view.
    // Fetch the devices from persistent data store
    NSManagedObjectContext *managedObjectContext = [self managedObjectContext];
    NSFetchRequest *fetchRequest = [[NSFetchRequest alloc] initWithEntityName:@"List"];
    self.devices = [[managedObjectContext executeFetchRequest:fetchRequest error:nil] mutableCopy];
    
    [self.tableView reloadData];
}


- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
   return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.devices.count;
}


- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    static NSString *CellIdentifier = @"Cell";
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:CellIdentifier forIndexPath:indexPath];
    
    // Configure the cell...
    NSManagedObject *device = [self.devices objectAtIndex:indexPath.row];
    [cell.textLabel setText:[NSString stringWithFormat:@"%@ - Age: %@", [device valueForKey:@"pet_name"], [device valueForKey:@"age"]]];
    
    cell.textLabel.numberOfLines = 1; // set the numberOfLines
    cell.textLabel.lineBreakMode = NSLineBreakByTruncatingTail;
    
    [cell.detailTextLabel setText:[NSString stringWithFormat:@"%@", [device valueForKey:@"date"]]];
    
    UIImage *image = [UIImage imageWithData:[device valueForKey:@"img"]];
    cell.imageView.image = image;
    
    return cell;
}



// Override to support conditional editing of the table view.
- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath {
    // Return NO if you do not want the specified item to be editable.
    return YES;
}


- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath
{
    NSManagedObjectContext *context = [self managedObjectContext];
    
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        // Delete object from database
        [context deleteObject:[self.devices objectAtIndex:indexPath.row]];
        
        NSError *error = nil;
        if (![context save:&error]) {
            NSLog(@"Can't Delete! %@ %@", error, [error localizedDescription]);
            return;
        }
        
        // Remove device from table view
        [self.devices removeObjectAtIndex:indexPath.row];
        [self.tableView deleteRowsAtIndexPaths:[NSArray arrayWithObject:indexPath] withRowAnimation:UITableViewRowAnimationFade];
    }
}
- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
     [self performSegueWithIdentifier:@"UpdateDevice" sender:self];
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    if ([[segue identifier] isEqualToString:@"UpdateDevice"]) {
        NSManagedObject *selectedDevice = [self.devices objectAtIndex:[[self.tableView indexPathForSelectedRow] row]];
        DeviceDetailViewController *destViewController = segue.destinationViewController;
        destViewController.device = selectedDevice;
    }
}

@end

File: Animal Age/DeviceViewController.swift
//
//  DeviceViewController.swift
//  Animal Age Converter
//
//  Created by Jon on 3/27/17.
//  Copyright © 2017 Jon Brown. All rights reserved.
//
import Foundation
import UIKit
import CoreData

class DeviceViewController: UITableViewController {
    private var devices: [AnyHashable]?

    func managedObjectContext() -> NSManagedObjectContext? {

        var context: NSManagedObjectContext? = nil
        let delegate = UIApplication.shared.delegate as? AppDelegate
        // call "persistentContainer" not "managedObjectContext"
        if delegate?.perform(#selector(getter: AppDelegate.persistentContainer)) != nil {
            // call viewContext from persistentContainer not "managedObjectContext"
            context = delegate?.persistentContainer?.viewContext
        }

        return context
    }

    override func viewDidLoad() {

        super.viewDidLoad()
        title = "Age Tracker"
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        // You code here to update the view.
        // Fetch the devices from persistent data store
        let managedObjectContext = self.managedObjectContext()
        let fetchRequest = NSFetchRequest<NSFetchRequestResult>(entityName: "List")
        do {
            devices = try managedObjectContext?.fetch(fetchRequest) as? [AnyHashable]
        } catch {
        }

        tableView.reloadData()
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

// MARK: - Table view data source
    override func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return devices?.count ?? 0
    }

    static let tableViewCellIdentifier = "Cell"

    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: DeviceViewController.tableViewCellIdentifier, for: indexPath)
        
       

        // Configure the cell...
        let device = devices?[indexPath.row] as? NSManagedObject
        if let value = device?.value(forKey: "pet_name") {
            cell.textLabel?.text = "\(value)"
        }

        cell.textLabel?.numberOfLines = 0 // set the numberOfLines
        cell.textLabel?.font = UIFont.boldSystemFont(ofSize: 16.0)
        cell.textLabel?.lineBreakMode = .byWordWrapping

        if let value = device?.value(forKey: "age") {
            cell.detailTextLabel?.text = "Age: \(value)"
        }

        var image: UIImage? = nil
        if let value = device?.value(forKey: "img") as? Data {
            image = UIImage(data: value)
        }
        
        let cellImageLayer: CALayer?  = cell.imageView?.layer
        
        cell.imageView?.image = image
        
        image = resizeImage(image: image!, toTheSize: CGSize(width: 70, height: 70))

       

       cellImageLayer!.cornerRadius = 35
       cellImageLayer!.masksToBounds = true
        
        cell.imageView?.image = image
        
        cell.frame.inset(by: UIEdgeInsets(top: 10, left: 10, bottom: 10, right: 10))

        return cell
    }

    
    // Override to support conditional editing of the table view.
    override func tableView(_ tableView: UITableView, canEditRowAt indexPath: IndexPath) -> Bool {
        // Return NO if you do not want the specified item to be editable.
        return true
    }

    override func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCell.EditingStyle, forRowAt indexPath: IndexPath) {
        let context = managedObjectContext()

        if editingStyle == .delete {
            // Delete object from database
            if let object = devices?[indexPath.row] as? NSManagedObject {
                context?.delete(object)
            }

            let error: Error? = nil
            do {
                try context?.save()

                if try context?.save() == nil {
                    if let error = error {
                        print("Can't Delete! \(error) \(error.localizedDescription )")
                    }
                    return
                }
            } catch {
                
                return
            }

            // Remove device from table view
            devices?.remove(at: indexPath.row)
            self.tableView.deleteRows(at: [indexPath], with: .fade)
        }
    }
    
    func resizeImage(image:UIImage, toTheSize size:CGSize)->UIImage{

        let scale = CGFloat(max(size.width/image.size.width,
                                size.height/image.size.height))
        let width:CGFloat  = image.size.width * scale
        let height:CGFloat = image.size.height * scale;

        let rr:CGRect = CGRect(x: 0, y: 0, width: width, height: height)

        UIGraphicsBeginImageContextWithOptions(size, false, 0);
        image.draw(in: rr)
        let newImage = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext();
        return newImage!
    }

    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        performSegue(withIdentifier: "UpdateDevice", sender: self)
    }

    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if (segue.identifier == "UpdateDevice") {
            let selectedDevice = devices?[tableView.indexPathForSelectedRow?.row ?? 0] as? NSManagedObject
            let destViewController = segue.destination as? DeviceDetailViewController
            destViewController?.device = selectedDevice
        }
    }
}
